### PR TITLE
resolve chapter 4

### PR DESCRIPTION
**Pull request by krivtsov**
### Solutions for Chapter 4

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter4.hs [hunk 1]
```diff
@@ -472,10 +478,11 @@ Implement the Applicative instance for our 'Secret' data type from before.
 -}
 instance Applicative (Secret e) where
     pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+    pure a = Reward a
 
     (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+    (Trap e) <*> _ = Trap e
+    (Reward f) <*> a = fmap f a
```
Comment (chshersh) [inline src/Chapter4.hs:485]: Nice work on reusing `Functor` instance for `Secret`! :world_map:

File: src/Chapter4.hs [hunk 2]
```diff
@@ -489,6 +496,17 @@ Implement the 'Applicative' instance for our 'List' type.
   type.
 -}
 
+append :: List a -> List a -> List a
+append Empty as = as
+append (Cons car cdr) l = Cons car (append cdr l)
```
Comment (chshersh) [inline src/Chapter4.hs:501]: I see a Lisp developer here 😄

**Review by chshersh — APPROVED** [review]
:tada: Fantastic work!
Thank you so much for taking this course! I hope you had fun :man_dancing: :microphone: 

We would really appreciate your [feedback on the course](https://docs.google.com/forms/d/e/1FAIpQLScBVhLxq5CgGnAfIGUE-fCoOUqeGkDY2HXzbT7KV2jjLOsmjQ/viewform)!

You can also ask any questions and we can discuss any topics you would like to dig into as well  :slightly_smiling_face: Don't be a stranger!

Meanwhile, do you need any guidance or hints regarding the advanced tasks? 🙂

**Comment by vrom911** [comment]
Thanks for participating in Learn4Haskell! Hope you enjoyed this Hacktoberfest 🥳 

I am going to close this PR as we can't merge the solutions into our course 🔐 

Even though the Hacktoberfest is over, we are going to continue reviewing, so feel free to open new PRs, if you haven't finished the course yet! 🎊

**Pull request closed without merging**